system/lib/wasmfs/file_table.h

```diff
@@ -19,28 +19,30 @@ namespace wasmfs {
 
 std::shared_ptr<Directory> getRootDirectory();
 
-class OpenFileState {
+class OpenFileState : public std::enable_shared_from_this<OpenFileState> {
   std::shared_ptr<File> file;
   __wasi_filedelta_t offset;
-  // An OpenFileState needs a mutex if there are concurrent accesses on one open file
-  // descriptor. This could occur if there are multiple seeks on the same open file descriptor.
+  // An OpenFileState needs a mutex if there are concurrent accesses on one open
+  // file descriptor. This could occur if there are multiple seeks on the same
+  // open file descriptor.
   std::mutex mutex;
 
 public:
-  OpenFileState(uint32_t offset, std::shared_ptr<File> file) : offset(offset), file(file) {}
+  OpenFileState(uint32_t offset, std::shared_ptr<File> file)
+    : offset(offset), file(file) {}
 
   class Handle {
-    OpenFileState& openFileState;
+    std::shared_ptr<OpenFileState> openFileState;
     std::unique_lock<std::mutex> lock;
 
   public:
-    Handle(OpenFileState& openFileState)
-      : openFileState(openFileState), lock(openFileState.mutex) {}
+    Handle(std::shared_ptr<OpenFileState> openFileState)
+      : openFileState(openFileState), lock(openFileState->mutex) {}
 
-    std::shared_ptr<File>& getFile() { return openFileState.file; };
+    std::shared_ptr<File>& getFile() { return openFileState->file; };
```

Comment:
Shouldn't this be just `return openFileState;`?

Reply:
Hmm, I think this is supposed to be returning the underlying `shared_ptr<File>` associated with the `openFileState`?